Refactor ci tests to improve speed.

jwithers opened the pull request:
This refactor reduces ci test time for all the tests in ci.yml to 16 to 19 minutes depending on weather. 

So our total time on pre-merge ci is going to be ~22 minutes. See bug for more details on runtimes of our current suite. We can get that down to sub 20 minutes if we can move the hifi5 xtensa test to pot-merge which is part of https://github.com/tensorflow/tflite-micro/pull/1772.

This PR will not pass our branch protection automation since it makes substantial changes to our tests. Additionally, the branch protection rules will need to change once this PR is merged.

BUG=https://issuetracker.google.com/issues/266654467

jwithers commented:
Have moved Hifi5 Unit Tests to post merge tests. This should bring our total presubmit test suite down to 16-19 minutes.

Also must be pulled from the branch protections:
REMOVE: Tests Entry Point / call-xtensa / Hifi5 Unit Tests (presubmit)

advaitjain commented on .github/workflows/ci.yml:
```diff
@@ -40,15 +40,115 @@ jobs:
           tensorflow/lite/micro/tools/ci_build/test_bazel.sh
 
-  cortex_m_tests:
+  bazel_tests_nooss:
```
nit: nooss --> no_oss

advaitjain commented on .github/workflows/ci.yml:
```diff
@@ -40,15 +40,115 @@ jobs:
     runs-on: ubuntu-latest
 
-    name: Cortex-M (presubmit)
+    name: Bazel nooss (presubmit)
```
nit: nooss --> no_oss

advaitjain commented on .github/workflows/ci.yml:
```diff
@@ -40,15 +40,115 @@ jobs:
+          pip3 install Wave
+      - name: Test
+        run: |
+          tensorflow/lite/micro/tools/ci_build/test_bazel_nooss.sh
```
nit: nooss --> no_oss

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_bazel.sh:
```diff
@@ -22,32 +22,5 @@ cd "${ROOT_DIR}"
 
 source tensorflow/lite/micro/tools/ci_build/helper_functions.sh
 
-# We are using a bazel build followed by bazel test to make sure that the CI
```
let's call this file `test_bazel_tflite_tools.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_bazel_nooss.sh:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
let's call this file `test_bazel.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_bluepill_no_release.sh:
```diff
@@ -33,17 +33,9 @@ TARGET=bluepill
 # TODO(b/143715361): downloading first to allow for parallel builds.
```
let's call this file `test_bluepill_default.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_bluepill_renode.sh:
```diff
@@ -0,0 +1,39 @@
+# We use Renode differently when running the full test suite (make test) vs an
+# individual test. So, we test only of the kernels individually as well to have
+# both of the Renode variations be part of the CI.
+readable_run make -j8 -f ${TENSORFLOW_ROOT}tensorflow/lite/micro/tools/make/Makefile TARGET=${TARGET} test_kernel_add_test TENSORFLOW_ROOT=${TENSORFLOW_ROOT} EXTERNAL_DIR=${EXTERNAL_DIR}
```
this line (`make test_kernel_add_test`) would be better included in `test_bluepill_default.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_stm32f4_no_release.sh:
```diff
@@ -32,11 +32,7 @@ readable_run make -f ${TENSORFLOW_ROOT}tensorflow/lite/micro/tools/make/Makefile
 # TODO(b/143715361): downloading first to allow for parallel builds.
 readable_run make -f ${TENSORFLOW_ROOT}tensorflow/lite/micro/tools/make/Makefile OPTIMIZED_KERNEL_DIR=${OPTIMIZED_KERNEL_DIR} TARGET=${TARGET} third_party_downloads TENSORFLOW_ROOT=${TENSORFLOW_ROOT} EXTERNAL_DIR=${EXTERNAL_DIR}
```
`test_stm32f4_default.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_x86_example.sh:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
```
let's call this `test_x86_out_of_tree.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_x86_no_release.sh:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
`test_x86_default.sh`

advaitjain commented on tensorflow/lite/micro/tools/ci_build/test_x86_no_static.sh:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
`test_x86_no_tflite_static_memory.sh`

advaitjain commented on .github/workflows/tests_post.yml:
```diff
@@ -0,0 +1,72 @@
+  f1_unit_tests:
+    runs-on: ubuntu-latest
+
+    name: Fusion F1 Unit Tests (presubmit)
```
presubmit --> postmerge

advaitjain commented on .github/workflows/tests_post.yml:
```diff
@@ -0,0 +1,72 @@
+  hifi5_unit_tests: 
+    runs-on: ubuntu-latest
+
+    name: Hifi5 Unit Tests (presubmit)
```
presubmit --> postmerge

jwithers commented:
Moved postmerge machinery to https://github.com/tensorflow/tflite-micro/pull/1772 and updated description accordingly. Changed test names as per @advaitjain comments and updated the branch protection test changes in the description accordingly. Have tested changes on fork and they work.

advaitjain commented:
> Have moved Hifi5 Unit Tests to post merge tests. This should bring our total presubmit test suite down to 16-19 minutes.
> 
> Also must be pulled from the branch protections: REMOVE: Tests Entry Point / call-xtensa / Hifi5 Unit Tests (presubmit)

This change is now in a separate PR https://github.com/tensorflow/tflite-micro/pull/1772

advaitjain reviewed (approved):
approving and will then merge bypassing the branch protection rules.

advaitjain commented:
I have updated the branch protection rules and https://github.com/tensorflow/tflite-micro/pull/1777 should be the first PR that is merged with the new workflows/jobs.

Also, created http://b/270657519 to figure out how many of these jobs can be moved to post-submit once the basic hooks for it are merged with https://github.com/tensorflow/tflite-micro/pull/1772
